backend/src/apiserver/resource/resource_manager.go

```diff
@@ -887,9 +888,30 @@ func (r *ResourceManager) ReportWorkflowResource(ctx context.Context, workflow *
 	if jobId == "" {
 		// If a run doesn't have job ID, it's a one-time run created by Pipeline API server.
 		// In this case the DB entry should already been created when argo workflow CR is created.
-		err := r.runStore.UpdateRun(runId, condition, workflow.FinishedAt(), workflow.ToStringForStore())
-		if err != nil {
-			return util.Wrap(err, "Failed to update the run.")
+		updateError := r.runStore.UpdateRun(runId, condition, workflow.FinishedAt(), workflow.ToStringForStore())
+		if updateError != nil {
+			if !util.IsUserErrorCodeMatch(updateError, codes.NotFound) {
+				return util.Wrap(updateError, "Failed to update the run.")
+			}
+			// Handle run not found in run store error.
+			// To avoid letting the workflow leak for ever, we need to GC it when its record does not exist in KFP DB.
+			glog.Errorf("Cannot find reported workflow name=%q namespace=%q runId=%q in run store. "+
+				"Deleting the workflow to avoid resource leaking. "+
+				"This can be caused by installing two KFP instances that try to manage the same workflows "+
+				"or an unknown bug. If you encounter this, recommend reporting more details in https://github.com/kubeflow/pipelines/issues/6189.",
+				workflow.GetName(), workflow.GetNamespace(), runId)
+			err := r.getWorkflowClient(workflow.Namespace).Delete(ctx, workflow.Name, v1.DeleteOptions{})
+			if err != nil {
+				if util.IsNotFound(err) {
+					return util.NewNotFoundError(err, "Failed to delete the obsolete workflow for run %s", runId)
```

Comment:
Do we need to throw error here? Can we just throw a warning log here?

Reply:
persistence agent calls api server, if there's an error and we do not return it.
persistence agent will think the sync is successful and remove the item from the queue.

In this case, there's no difference between returning a not found error and silently log a warning and return successful, because when persistence agent gets a not found error, it also remove the item from the queue.

I'm slightly leaning towards returning the error here, because I think for reportWorkflow, it should just do its job and return whatever error it encountered. Then the persistence agent that calls report workflow should use the return status code to decide what to do with the error.

I wonder if that makes sense to you

Reply:
it makes sense to me